Blog: Using Admission Controllers to Detect Container Drift at Runtime (#29761)
kubernetes/website: content/en/blog/_posts/2021-09-20-admission-controllers-and-container-drift/index.md

### 1.  Receive the admission request of interactive kubectl command
We need to first define qualified admission webhooks to receive specific requests in our admission controller service. In the interest of `kubectl exec`, for example, we configure the webhook’s rules with resources being “pods/exec” and operations as “CONNECT”. They tell the K8s API server that all “exec” requests should be subject to our admission controller service. We also specify a “url” which gives the location of our service and “caBundle” to provide its certificate, both under the “clientConfig” stanza. Here is a short example of what our ValidatingWebhookConfiguration object looks like:
```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: "example-webhook"
webhooks:
- name: "admit-pod-exec-webhook"
  rules:
  - apiGroups: ["*"]
    apiVersions: ["*"]
    operations: ["CONNECT"]
    resources: ["pods/exec"]
  clientConfig:
    url: "https://example:9443/admit-pod-exec"
    caBundle: "<PEM encoded CA bundle>"
```

### 2.  Identify the pod with potentially mutated container

> Once a request of `kubectl exec` comes to our admission controller service, it identifies the target pod by adding a customized K8s label. In this way, we can not only query all the pods with potentially mutated containers, but also enable the service to retrieve previously identified pods in case it gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object given in the admission request is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the label by a Kube client connecting to the cluster. The service always returns an admission response as allowed after passing that pod reference to its Kube client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected simply by running a `kubectl describe` command:

Reviewer: ```suggestion
Once a request of `kubectl exec` comes to our admission controller service, it identifies the target pod by adding a customized K8s label. In this way, we can not only query all the pods with potentially mutated containers, but also enable the service to retrieve previously identified pods in case it gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object given in the admission request is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the label by a Kube client connecting to the cluster. The service always returns an admission response as allowed after passing that pod reference to its Kube client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected or not, simply by running a `kubectl describe` command:
```
A general suggestion you might want to include.

{{< figure src="workflow-diagram.png" alt="Workflow Diagram" >}}

## New kubectl plugin for a better user experience
Our admission controller service works great for solving the container drift issue we had on the platform. It also submits related K8s events to the target pod that has been affected. However, as K8s events get away quickly (one hour of persistency by default configuration), we need to provide other mechanisms for developers to get their pod interaction activity. A [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) is a perfect choice for us to expose this information. We named our plugin “kubectl-pi” (short for pod-interaction) and provided two subcommands: “get” and “extend”. When the “get” command is called, the plugin simply checks the label attached by our admission controller service and transfers it to human-readable info. Here is an example output of the `kubectl pi get` command:

```
$ kubectl pi get test-pod
POD_NAME    INTERACTOR    POD_TTL    EXTENSION    EXTENSION_REQUESTER    EVICTION_TIME
test-pod    username-1    1h0m0s     /            /                      2021-09-20 11:00:00 -0800 PST
```

> The plugin can also be used to extend the current eviction time of a pod, in case developers need more buffer to debug ongoing issues after exec’ing into their containers. This is achieved by the `kubectl pi extend` command, where the plugin patches relevant K8s annotation to the given pod. This annotation includes the duration and username who made the extension request for transparency (displayed in the table returned from “get” command). There is another webhook defined in our admission controller service that admits this annotation request and resets the eviction timer of the target pod accordingly. An example of requesting the extension from the developer side would be:

Reviewer: ```suggestion
The plugin can also be used to extend the current eviction time of a pod, in case developers need more buffer to debug ongoing issues after executing on their containers. This is achieved by the `kubectl pi extend` command, where the plugin patches relevant K8s annotation to the given pod. This annotation includes the duration and username who made the extension request for transparency (displayed in the table returned from “get” command). There is another webhook defined in our admission controller service that admits this annotation request and resets the eviction timer of the target pod accordingly. An example of requesting the extension from the developer side would be:
```

---
layout: blog
title: "Using Admission Controllers to Detect Container Drift at Runtime"
date: 2021-09-20
slug: admission-controllers-and-container-drift
---

**Author:** Saifuding Diliyaer (Box)
{{< figure src="intro-illustration.png" alt="Intro Illustration" width="70%" >}}

> *Illustration by Munire Aireti*

Reviewer 2: nit: you can put the attribution into the figure

See https://gohugo.io/content-management/shortcodes/#figure for details.

Author: @sftim Thanks for the tip. Updated the PR with your suggestion.
